include/crpropa/PhotonBackground.h

```diff
@@ -2,36 +2,214 @@
 #define CRPROPA_PHOTONBACKGROUND_H
 
 #include "crpropa/Common.h"
+#include "crpropa/Referenced.h"
 
+#include <vector>
 #include <string>
 
 namespace crpropa {
-
 /**
- * \addtogroup EnergyLosses
+ * \addtogroup PhotonFields
  * @{
  */
-// Photon fields
-// The default IRB model is that of Kneiske et al. 2004
-enum PhotonField {
-	CMB,
-	IRB,  // same as IRB_Kneiske04
-	IRB_Kneiske04,
-	IRB_Stecker05,
-	IRB_Franceschini08,
-	IRB_Finke10,
-	IRB_Dominguez11,
-	IRB_Gilmore12,
-	IRB_Stecker16_upper,
-	IRB_Stecker16_lower,
-	URB_Protheroe96
+
+/**
+ @class PhotonField
+ @brief Abstract base class for photon fields.
+ */
+class PhotonField: public Referenced {
+public:
+	PhotonField() {
+		this->fieldName = "AbstractPhotonField";
+		this->isRedshiftDependent = false;
+	}
+
+	/**
+	 returns comoving photon density [1/m^3].
+	 multiply with (1+z^3) for physical number density.
+	 @param ePhoton		photon energy [J]
+	 @param z			redshift (if redshift dependent, default = 0.)
+	 */
+	virtual double getPhotonDensity(double ePhoton, double z = 0.) const = 0;
+
+	/**
+	 returns overall comoving scaling factor
+	 (cf. CRPropa3-data/calc_scaling.py)
+	 @param z		redshift
+	 */
+	virtual double getRedshiftScaling(double z) const {
+		return 1.;
+	};
+
+	bool hasRedshiftDependence() const {
+		return this->isRedshiftDependent;
+	}
+
+	std::string getFieldName() const {
+		return this->fieldName;
+	}
+
+protected:
+	std::string fieldName;
+	bool isRedshiftDependent;
+};
+
+/**
+ @class TabularPhotonField
+ @brief Photon field decorator for tabulated photon fields.
+
+ This class reads photon field data from files;
+ The first file must be a list of photon energies [J], named fieldName_photonEnergy.txt
+ The second file must be a list of comoving photon field densities [1/m^3], named fieldName_photonDensity.txt
+ Optionally, a third file contains redshifts, named fieldName_redshift.txt
+ */
+class TabularPhotonField: public PhotonField {
+public:
+	TabularPhotonField(const std::string fieldName, const bool isRedshiftDependent = true);
+	double getPhotonDensity(double ePhoton, double z = 0.) const;
+	double getRedshiftScaling(double z) const;
+
+protected:
+	void readPhotonEnergy(std::string filePath);
+	void readPhotonDensity(std::string filePath);
+	void readRedshift(std::string filePath);
+	void initRedshiftScaling();
+	void checkInputData() const;
+
+	std::vector<double> photonEnergies;
+	std::vector<double> photonDensity;
+	std::vector<double> redshifts;
+	std::vector<double> redshiftScalings;
+};
+
+/**
+ @class IRB_Kneiske04
+ @brief Extragalactic background light model from Kneiske et al. 2004
+
+ Source info:
+ DOI:10.1051/0004-6361:20031542,
+ https://www.aanda.org/articles/aa/pdf/2004/03/aa3848.pdf, figure 1 (”Best-fit” model)
```

Comment:
@Froehliche-Kernschmelze Please change quotation marks here into regular (ASCII) ones, i.e. to "Best-fit". Otherwise, it seems that we cannot generate documentation from this comment (in TravisCI). Then commit.